### PR TITLE
Add dependency on net-http gem

### PR DESCRIPTION
**Pull request by stanhu**
As described in https://www.ruby-lang.org/en/news/2020/12/25/ruby-3-0-0-released/, Ruby
3.0.0 promoted net-http from stdlib. Adding this as a dependency avoids
"already initialized constant errors" if net-imap or some other gem is
included. See https://github.com/ruby/net-imap/issues/16#issuecomment-803086765 for
more details.

### CODE REVIEW

File: faraday-net_http.gemspec [hunk 1]
```diff
@@ -22,6 +22,8 @@ Gem::Specification.new do |spec|
   spec.files = Dir.glob('lib/**/*') + %w[README.md LICENSE.md]
   spec.require_paths = ['lib']
 
+  spec.add_dependency 'net-http', '~> 0.1'
```
Comment (stanhu) [inline faraday-net_http.gemspec:25]: This needs Ruby 2.6.0. I suppose we can just demote this to a development dependency until Ruby 2.5 is no longer supported?

**Comment by rizalmuthi-ap** [comment]
Could we get this one merge please.
Thank you @stanhu @JanDintel @iMacTia @olleolleolle @tricknotes

**Comment by iMacTia** [comment]
Thanks @stanhu for bringing this to our attention and apologies for the delay in getting back! For some reason this repo was not on my watchlist, so I was only notified thanks to @rizalmuthi-ap comment 😅

I had a quick look and I'm not sure on how to progress with this, because it seems to me as if it's impossible to make this work "out-of-the-box" without removing support for Ruby 2.5:
* If we simply add the dependency to the `gemspec`, it will work just fine, but it will break Ruby 2.5
* If we add it as a development dependency, I believe it will "break" again on production deployments?
* If we do nothing people encountering this issue can still fix it adding `net-http` to their gemfile

Since I'd like to keep supporting Ruby 2.5, it's hard to pick one or the other option.
To add more on that, I understand this issue only affects Ruby 3.0 users with `net-imap` on their Gemfile, and to be honest I'm also surprised we have this issue in the first place 🤔. Looking at the linked comment:

> * When the faraday-net_http gem is required, it then requires net/http.
> * Then the stdlib version of net/http is loaded, since bundler doesn't know net-http is a gem dependency, so it doesn't add the gemified version to to the $LOAD_PATH.

I'm wonderig why that's the case? Why would the stdlib version of the library be loaded when you're running Ruby 3? Shouldn't ruby be smart enough to know that for versions <3 it needs to load the stdlib one, and for version 3+ the default gem? Why would they push this responsibility to the developer? It doesn't make sense honestly.

Looking at how [default gems work](https://nts.strzibny.name/ruby-stdlib-default-bundled-gems/) I'm surprised we're having this issue at all and I suspect it might be due to the fact that `net-imap` is explicitly added to the Gemfile. That should not be necessary from what I understand, because that's also a default gem? Why would that be necessary?

In short: I believe this is outside of the responsibility of this library and we should delegate to Ruby the selection of the right library. If the developer needs to add `net-imap` for whatever reason, then they can simply add `net-http` as well to fix the warnings.
I'd prefer not to drop Ruby 2.5 compatibility to solve what seems like an unlikely scenario.

Sorry for the long comment, most of this was new to me so I might have misunderstood something 😄. Feel free to point that out if that's the case

**Comment by ClearlyClaire** [comment]
> I'm wonderig why that's the case? Why would the stdlib version of the library be loaded when you're running Ruby 3? Shouldn't ruby be smart enough to know that for versions <3 it needs to load the stdlib one, and for version 3+ the default gem? Why would they push this responsibility to the developer? It doesn't make sense honestly.

From what I understand the issue does not occur on Ruby 3 (where the code is loaded from the `net-http` gem, whether it's implicit or not) but on earlier versions, and the issue occurs when the project explicitly depends on the `net-protocol` gem (either directly or through another dependency) and `faraday-net_http` without also explicitly depending on `net-http`, in which case `faraday-net_http` will load the `net/http` code and thus the `net/protocol` code from Ruby's stdlib and *also* load the `net-protocol` gem.

I'm not sure where the responsibility stands here, I'd say if `net-protocol` is a default gem, maybe it should not get explicitly required by the project (or its dependencies) in the first place. Explicitly depending on `net-http` as in this PR would also solve the issue but at the cost of Ruby 2.5 support.

**Comment by iMacTia** [comment]
The issue here is that `mail_room` has explicitly [added `net-imap` to their gemspec](https://github.com/tpitale/mail_room/pull/119), impacting the default behaviour of using the internal stdlib in older ruby versions and breaking compatibility with Ruby 2.5.

I don't think explicitly adding the dependency on net-http is a good idea here, first because we'd drop compatibility with Ruby 2.5, and second because we might cause similar issues upstream for other gems. There's also the fact that Faraday is a `development_dependency` for `mail_room`.

What I think should happen instead, is any of the following:

* `mail_room` should remove `net-imap` from their `gemspec`
* `mail_room` should add `net-http` as a `development_dependency`
* Ruby should be smart enough to load the gemified version when available, rather than the stdlib 🤔 (still surprised this is not the case already)

**Comment by ClearlyClaire** [comment]
Note that in Mastodon we had a similar issue with a dependency pulling `net-ftp`, and `faraday-net_http` using the stdlib `net/http`. This has been resolved for us by our dependency dropping `net-ftp`.

**Comment by deivid-rodriguez** [comment]
> > * When the faraday-net_http gem is required, it then requires net/http.
> > * Then the stdlib version of net/http is loaded, since bundler doesn't know net-http is a gem dependency, so it doesn't add the gemified version to to the $LOAD_PATH.

> I'm wonderig why that's the case? Why would the stdlib version of the library be loaded when you're running Ruby 3? Shouldn't ruby be smart enough to know that for versions <3 it needs to load the stdlib one, and for version 3+ the default gem? Why would they push this responsibility to the developer? It doesn't make sense honestly.

I'm not sure I understand what you think it's wrong, since I think your expectation:

> Shouldn't ruby be smart enough to know that for versions <3 it needs to load the stdlib one, and for version 3+ the default gem?

is exactly what should be happening. When `net-http` is a default gem, `require` will activate the latest version of the gem present in the system. When `net-http` is not a default gem, `require` will load the stdlib version.

**Comment by iMacTia** [comment]
I mean, if it was that simple we wouldn't have this double-load issue would we 😅 .

I guess my wish would be that Ruby would be smart enough to understand that if I have a gem specified in my Gemfile or any of my dependencies, it should load that one. If there's no gem specified, then the stdlib/standard gem is fine.

Right now it's a mess because it load one AND the other under certain circumstances
